File: src/main/scala/com/twitter/joauth/Verifier.scala
// Copyright 2011 Twitter, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
// file except in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

package com.twitter.joauth

import org.slf4j.LoggerFactory

/**
 * A Verifier takes either
 *   a) an OAuth1 request, a token secret, and a consumer secret
 *      or
 *   b) an OAuth1TwoLegged request and a consumer secret
 *
 * and validates the request. It returns a Java enum for compatability
 */
trait Verifier {
  def apply(request: OAuth1Request, tokenSecret: String, consumerSecret: String): VerifierResult
  def apply(request: OAuth1TwoLeggedRequest, consumerSecret: String): VerifierResult
}

/**
 * for testing. always returns the same result.
 */
class ConstVerifier(result: VerifierResult) extends Verifier {
  override def apply(request: OAuth1Request, tokenSecret: String, consumerSecret: String): VerifierResult = result
  override def apply(request: OAuth1TwoLeggedRequest, consumerSecret: String): VerifierResult = result
}

/**
 * a factory with various convenience constructors for a StandardVerifier
 */
object Verifier {
  val NO_TIMESTAMP_CHECK = -1

  def apply(): Verifier = new StandardVerifier(
    Signer(), NO_TIMESTAMP_CHECK, NO_TIMESTAMP_CHECK, NoopNonceValidator)
  def apply(maxClockFloatAheadMins: Int, maxClockFloatBehindMins: Int) = new StandardVerifier(
    Signer(), maxClockFloatAheadMins, maxClockFloatBehindMins, NoopNonceValidator)
  def apply(
    maxClockFloatAheadMins: Int,
    maxClockFloatBehindMins: Int,
    validateNonce: NonceValidator) = new StandardVerifier(
    Signer(), maxClockFloatAheadMins, maxClockFloatBehindMins, validateNonce)
  def apply(
    sign: Signer,
    maxClockFloatAheadMins: Int,
    maxClockFloatBehindMins: Int,
    validateNonce: NonceValidator) = new StandardVerifier(
    sign, maxClockFloatAheadMins, maxClockFloatBehindMins, validateNonce)
}

/**
 * The standard implementation of a Verifier. Constructed with a Signer, the maximum clock float
 * allowed for a timestamp, and a NonceValidator.
 */
object StandardVerifier {
  private val log = LoggerFactory.getLogger(getClass.getName)
}

class StandardVerifier(
  signer: Signer,
  maxClockFloatAheadMins: Int,
  maxClockFloatBehindMins: Int,
  validateNonce: NonceValidator)
extends Verifier {

  import StandardVerifier._

  val maxClockFloatAheadSecs = maxClockFloatAheadMins * 60L
  val maxClockFloatBehindSecs = maxClockFloatBehindMins * 60L

  override def apply(request: OAuth1Request, tokenSecret: String, consumerSecret: String): VerifierResult = {
    verifyOAuth1(
      request,
      request.nonce,
      request.timestampSecs,
      tokenSecret,
      consumerSecret,
      request.signature,
      request.normalizedRequest
    )
  }

  override def apply(request: OAuth1TwoLeggedRequest, consumerSecret: String): VerifierResult = {
    verifyOAuth1(
      request,
      request.nonce,
      request.timestampSecs,
      "",
      consumerSecret,
      request.signature,
      request.normalizedRequest
    )
  }

  private def verifyOAuth1(
    request: OAuthRequest,
    nonce: String,
    timestampSecs: Long,
    tokenSecret: String,
    consumerSecret: String,
    signature: String,
    normalizedRequest: String) = {
    if (!validateTimestampSecs(timestampSecs)) {
      log.debug("bad timestamp -> {}", request.toString)
      VerifierResult.BAD_TIMESTAMP
    } else if (!validateNonce(nonce)) {
      log.debug("bad nonce -> {}", request.toString)
      VerifierResult.BAD_NONCE
    } else if (!validateSignature(normalizedRequest, signature, tokenSecret, consumerSecret)) {
      log.debug("bad signature -> {}", request.toString)
      VerifierResult.BAD_SIGNATURE
    }
    else VerifierResult.OK
  }

  def validateTimestampSecs(timestampSecs: Long): Boolean = {
    val nowSecs = System.currentTimeMillis / 1000
    (maxClockFloatBehindMins < 0 || (timestampSecs >= nowSecs - maxClockFloatBehindSecs)) &&
    (maxClockFloatAheadMins < 0 || (timestampSecs <= nowSecs + maxClockFloatAheadSecs))
  }

  def validateSignature(
    normalizedRequest: String,
    signature: String,
    tokenSecret: String,
    consumerSecret: String): Boolean = {
    try {
      Base64Util.equals(UrlDecoder(signature).trim,
        signer.getBytes(normalizedRequest, tokenSecret, consumerSecret))
    } catch {
      case e: Exception => return false
    }
  }
}

File: src/main/java/com/twitter/joauth/Verifier.java
// Copyright 2011 Twitter, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
// file except in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

package com.twitter.joauth;

import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * A Verifier takes either
 *   a) an OAuth1 request, a token secret, and a consumer secret
 *      or
 *   b) an OAuth1TwoLegged request and a consumer secret
 *
 * and validates the request. It returns a Java enum for compatability
 */
public interface Verifier {
  static int NO_TIMESTAMP_CHECK = -1;

  //TODO: swap the order of the consumerSecret and tokenSecret because consumer secrets are required for all oauth1 requests

  public VerifierResult verify(UnpackedRequest.OAuth1Request request, String tokenSecret, String consumerSecret);
  public VerifierResult verify(UnpackedRequest.OAuth1TwoLeggedRequest request, String consumerSecret);

  /**
   * a factory with various convenience constructors for a StandardVerifier
   */
  public static class VerifierFactory {

    public static Verifier newVerifier() {
      return newVerifier(
        Signer.getStandardSigner(),
        NO_TIMESTAMP_CHECK,
        NO_TIMESTAMP_CHECK,
        NonceValidator.NO_OP_NONCE_VALIDATOR
      );
    }

    public static Verifier newVerifier(int maxClockFloatAheadMins, int maxClockFloatBehindMins) {
      return newVerifier(
        Signer.getStandardSigner(),
        maxClockFloatAheadMins,
        maxClockFloatBehindMins,
        NonceValidator.NO_OP_NONCE_VALIDATOR
      );
    }

    public static Verifier newVerifier(
      int maxClockFloatAheadMins,
      int maxClockFloatBehindMins,
      NonceValidator validateNonce
    ) {
      return newVerifier(Signer.getStandardSigner(), maxClockFloatAheadMins, maxClockFloatBehindMins, validateNonce);
    }

    public static Verifier newVerifier(
      Signer sign,
      int maxClockFloatAheadMins,
      int maxClockFloatBehindMins,
      NonceValidator validateNonce
    ) {
      return new StandardVerifier(sign, maxClockFloatAheadMins, maxClockFloatBehindMins, validateNonce);
    }
  }


  public static class StandardVerifier implements Verifier {

    private final Signer signer;
    private final int maxClockFloatAheadMins;
    private final int maxClockFloatBehindMins;
    private final NonceValidator validateNonce;
    private final long maxClockFloatAheadSecs;
    private final long maxClockFloatBehindSecs;

    private static final Logger log = Logger.getLogger("StandardVerifier");

    public StandardVerifier(
      Signer signer,
      int maxClockFloatAheadMins,
      int maxClockFloatBehindMins,
      NonceValidator validateNonce
    ) {
      this.signer = signer;
      this.maxClockFloatAheadMins = maxClockFloatAheadMins;
      this.maxClockFloatBehindMins = maxClockFloatAheadMins;
      this.validateNonce = validateNonce;

      maxClockFloatAheadSecs = maxClockFloatAheadMins * 60L;
      maxClockFloatBehindSecs = maxClockFloatBehindMins * 60L;
    }

    @Override
    public VerifierResult verify(UnpackedRequest.OAuth1TwoLeggedRequest request, String consumerSecret) {
      return verifyOAuth1(
        request,
        request.nonce(),
        request.timestampSecs(),
        "",
        consumerSecret,
        request.signature(),
        request.normalizedRequest()
      );
    }

    @Override
    public VerifierResult verify(UnpackedRequest.OAuth1Request request, String tokenSecret, String consumerSecret) {
      return verifyOAuth1(
        request,
        request.nonce(),
        request.timestampSecs(),
        tokenSecret,
        consumerSecret,
        request.signature(),
        request.normalizedRequest()
      );
    }

    private VerifierResult verifyOAuth1(
      UnpackedRequest.OAuthRequest request,
      String nonce,
      long timestampSecs,
      String tokenSecret,
      String consumerSecret,
      String signature,
      String normalizedRequest
    ) {
      if (!validateTimestampSecs(timestampSecs)) {
        if (log.isLoggable(Level.FINE)) {
          log.log(Level.FINE, String.format("bad timestamp -> %s", request.toString()));
        }
        return VerifierResult.BAD_TIMESTAMP;
      } else if (!validateNonce.validate(nonce)) {
        if (log.isLoggable(Level.FINE)) {
          log.log(Level.FINE, String.format("bad nonce -> %s", request.toString()));
        }
        return VerifierResult.BAD_NONCE;
      } else if (!validateSignature(normalizedRequest, signature, tokenSecret, consumerSecret)) {
        if (log.isLoggable(Level.FINE)) {
          log.log(Level.FINE, String.format("bad signature -> %s", request.toString()));
        }
        return VerifierResult.BAD_SIGNATURE;
      } else {
        return VerifierResult.OK;
      }
    }

    public boolean validateTimestampSecs(long timestampSecs) {
      long nowSecs = System.currentTimeMillis() / 1000;

      return (maxClockFloatBehindMins < 0 || (timestampSecs >= nowSecs - maxClockFloatBehindSecs)) &&
        (maxClockFloatAheadMins < 0 || (timestampSecs <= nowSecs + maxClockFloatAheadSecs));
    }

    boolean validateSignature(
      String normalizedRequest,
      String signature,
      String tokenSecret,
      String consumerSecret
    ) {
      try {
        return Base64Util.equals(UrlCodec.decode(signature).trim(),
            signer.getBytes(normalizedRequest, tokenSecret, consumerSecret));
      } catch (Exception e) {
        return false;
      }
    }
  }
}
